DOC: Move Script Repository to readthedocs

lassoan opened the pull request:
This commit adds content of Slicer Script Repository (https://www.slicer.org/wiki/Documentation/Nightly/ScriptRepository) as of 2021-04-22 to Slicer documentation.

This makes the example scripts versioned together with the Slicer source code and makes it editable using standard github contribution process.

Restructured text format is used, as this is the preferred file format for developer documentation (more complicated syntax than markdown, but developers should be familiar with it anyway and it offers more structured organization and formatting).

Created by converting from mediawiki to rst using pandoc, then manually fixing formatting errors (mostly links, code blocks, and figures).

lassoan commented:
Here is the generated preview of the page: https://slicer--5614.org.readthedocs.build/en/5614/developer_guide/script_repository.html

sjh26 commented:
👍

jcfr commented:
Thanks for working on this :pray: 

Few comments:

* integration in table of content

| This PR | Suggested change |
|--|--|
| ![image](https://user-images.githubusercontent.com/219043/115766134-9880a180-a375-11eb-98f1-ceedeeefdab0.png) | ![image](https://user-images.githubusercontent.com/219043/115765655-08daf300-a375-11eb-9789-c168937f121a.png) |



* update references to `slice.util`

Since the python API associated with the `slice.util` module is generated, I suggest to use the following whenever possible

```
:func:`slicer.util.getNode()`
```

* Group all example in categories. It looks likes we could have move existing example into `Markups`, and also introduce sub categories like `Viewers`, `Loading/Saving`, ...

jcfr commented on Docs/developer_guide/script_repository.rst:
```diff
@@ -0,0 +1,4237 @@
+
+Get a MRML node in the scene based on the node name and call methods of that object. For the MRHead sample data:
+
+.. code-block:: python
```
```suggestion
.. code-block:: python

```

jcfr replied:
This add missing line

jcfr commented on Docs/developer_guide/script_repository.rst:
```diff
@@ -0,0 +1,4237 @@
+
+-  How to open an .mrb file with Slicer at the command line?
+
+   Slicer.exe --python-code "slicer.util.loadScene( 'f:/2013-08-23-Scene.mrb' )"
```
```suggestion
::

   Slicer.exe --python-code "slicer.util.loadScene( 'f:/2013-08-23-Scene.mrb' )"
```

jcfr commented on Docs/developer_guide/script_repository.rst:
```diff
@@ -0,0 +1,4237 @@
+
+-  How to run a script in the Slicer environment in batch mode (without showing any graphical user interface)?
+
+   Slicer.exe --python-code "doSomething; doSomethingElse; etc." --testing --no-splash --no-main-window
```
```suggestion
::

   Slicer.exe --python-code "doSomething; doSomethingElse; etc." --testing --no-splash --no-main-window
```

jcfr commented on Docs/developer_guide/script_repository.rst:
```diff
@@ -0,0 +1,4237 @@
+   dicomDataDir = "c:/my/folder/with/dicom-files"  # input folder with DICOM files
+   loadedNodeIDs = []  # this list will contain the list of all loaded node IDs
+
+   from DICOMLib import DICOMUtils
```
To suggest best practices, I am wondering if we should move the import at the top

jcfr commented on Docs/developer_guide/script_repository.rst:
```diff
@@ -0,0 +1,4237 @@
+     from os.path import isfile, join
+     for f in listdir(incomingVolumeFolder):
+       if f in incomingVolumesProcessed:
+         # this is an incoming file, it was already there
```
```suggestion
         # This is an incoming file, it was already there
```

jcfr commented on Docs/developer_guide/script_repository.rst:
```diff
@@ -0,0 +1,4237 @@
+         continue
+       filePath = join(incomingVolumeFolder, f)
+       if not isfile(filePath):
+         # ignore directories
```
```suggestion
         # Ignore directories
```

jcfr replied:
For consistency, I suggest to have the first letter of each comment capitalized

jcfr commented on Docs/developer_guide/script_repository.rst:
```diff
@@ -0,0 +1,4237 @@
+
+.. code-block:: python
+
+volumeName='2: ENT IMRT'
```
```suggestion
   volumeName = '2: ENT IMRT'
```

jcfr commented on Docs/developer_guide/script_repository.rst:
```diff
@@ -0,0 +1,4237 @@
+       if not isfile(filePath):
+         # ignore directories
+         continue
+       logging.info("Loading new file: "+f)
```
```suggestion
       logging.info("Loading new file: " + f)
```

jcfr replied:
Add space around operators

jcfr commented on Docs/developer_guide/script_repository.rst:
```diff
@@ -0,0 +1,4237 @@
+
+.. code-block:: python
+
+   db=slicer.dicomDatabase
```
```suggestion
   db = slicer.dicomDatabase
```

jcfr commented on Docs/developer_guide/script_repository.rst:
```diff
@@ -0,0 +1,4237 @@
+
+volumeName='2: ENT IMRT'
+   n=slicer.util.getNode(volumeName)
+   instUids=n.GetAttribute('DICOM.instanceUIDs').split()
```
```suggestion
    instUids = n.GetAttribute("DICOM.instanceUIDs").split()
```

jcfr replied:
In addition of adding spaces around the `=`, I also suggest we also consistently use double quotes

jcfr commented on Docs/developer_guide/script_repository.rst:
```diff
@@ -0,0 +1,4237 @@
+How can I convert DICOM to NRRD on the command line?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+   /Applications/Slicer-4.6.2.app/Contents/MacOS/Slicer --no-main-window --python-code "node=slicer.util.loadVolume('/tmp/series/im0.dcm'); slicer.util.saveNode(node, '/tmp/output.nrrd'); exit()"
```
```suggestion
::

   /Applications/Slicer-4.6.2.app/Contents/MacOS/Slicer --no-main-window --python-code "node=slicer.util.loadVolume('/tmp/series/im0.dcm'); slicer.util.saveNode(node, '/tmp/output.nrrd'); exit()"
```

jcfr commented on Docs/developer_guide/script_repository.rst:
```diff
@@ -0,0 +1,4237 @@
+   for viewNode in viewNodes:
+     viewNode.SetSliceIntersectionVisibility(1)
+
+How to find similar functions? For this one I searched for "slice intersections" text in the whole slicer source code, found that the function is implemented in Base\QTGUI\qSlicerViewersToolBar.cxx, then translated the qSlicerViewersToolBarPrivate::setSliceIntersectionVisible(bool visible) method to Python.
```
```suggestion
  .. note::

     How to find similar functions?
     
     For this one I searched for "slice intersections" text in the whole slicer source code, found that the function is implemented in ``Base\QTGUI\qSlicerViewersToolBar.cxx``, then translated the ``qSlicerViewersToolBarPrivate::setSliceIntersectionVisible(bool visible)`` method to Python.
```

jcfr commented on Docs/developer_guide/script_repository.rst:
```diff
@@ -0,0 +1,4237 @@
+.. code-block:: python
+
+   # Update the sphere from the fiducial points
+   def UpdateSphere(param1, param2):
```
```suggestion
   def UpdateSphere(param1, param2):
     """Update the sphere from the fiducial points
     """
```

jcfr replied:
Use docstrings when it applies and remove comment

lassoan commented:
@jcfr you can take over now, I've committed all your suggestions.

jcfr commented:
There are still issue with some of heading not being rendered.

For example, in this section,  the `Get the pseudo-singleton subject hierarchy node` heading is not rendered:

![image](https://user-images.githubusercontent.com/219043/115786064-4f891700-a38e-11eb-8a51-e920a8620c25.png)

![image](https://user-images.githubusercontent.com/219043/115786132-662f6e00-a38e-11eb-9977-aad2efa44389.png)

lassoan commented:
Headers are missing because of inconsistent underline character usage.

We should use these: https://www.odoo.com/documentation/user/14.0/contributing/documentation/rst_cheat_sheet.html#headings

@jcfr do you prefer to fix it or you would rather pass it back to me?

jcfr commented:
I will let you take over the topic. I am done for the day. 🙏

lassoan commented:
I've fixed all the issues, categorized all the snippets, split them up to multiple source files (still one rendered page). Please have a final look: https://slicer--5614.org.readthedocs.build/en/5614/developer_guide/script_repository.html

I've kept the "Script repository" as it contains a few shell scripts and we'll can move C++ examples there, too (sphinx has nice multi-language code blocks feature).

jcfr commented:
The organization in categories looks great :100: 

Here is a suggestion to re-order them:
* Module specific categories are sorted alphabetically 

|  This PR |  Suggested |
|---|---|
| ![image](https://user-images.githubusercontent.com/219043/115930411-69426100-a457-11eb-8b36-3e2c6fa41a9d.png) | ![image](https://user-images.githubusercontent.com/219043/115930672-e372e580-a457-11eb-94fa-68e4d11e8e70.png)|

jcfr commented:
> I've kept the "Script repository" as it contains a few shell scripts and we'll can move C++ examples there, too (sphinx has nice multi-language code blocks feature).

Make sense

jcfr commented:
I am also wondering if moving: `Slice views`, `3D views`, `Common view operations` and `Access VTK rendering classes` under a `Views` category would be sensible ?
